### PR TITLE
feat(api): add aspirate handling to protocol engine

### LINKED ISSUE #6598: feat(robot-server): handle aspirate command in live protocol session
- [ ] Always aspirate from specific well
- [ ] Use HC to move pipette to aspirate location
- [ ] Use HC to aspirate liquid

### PR DESCRIPTION
**Pull request by mcous**
## Overview

Closes #6598

This PR adds aspirate logic to the Protocol Engine. While developing this guide, I also wrote up exactly _how_ I developed it as part of the new [Protocol Engine cookbook](https://coda.io/d/Protocol-Engine_dKsxOIJgBlG). See the [Adding a New Command](https://coda.io/d/Protocol-Engine_dKsxOIJgBlG/Adding-a-new-command_suUKz#_luyCF) section of the doc for:

- The order in which everything was implemented
    - TL;DR: outside in 
- How I decided what needed to be implemented
    - TL;DR: I used tests to discover units
- How I tested everything
    - TL;DR: [discovery testing](https://github.com/testdouble/contributing-tests/wiki/Discovery-Testing) as I worked down the dependency tree 

Roughly, this PR creates  and/or enhances the following classes and methods:

dependencies-6

## Changelog

The price is on the can

## Review requests

Please please please give the cookbook + corresponding write-up a read. This PR and the associated documentation is meant to start a more concerted effort around on-boarding devs into working on the Protocol Engine. I'm specifically looking for comments and edits to improve that documentation.

Otherwise, I've got a couple more specific items to discuss:

### Smoke testing

This PR **has not been tested on a robot**. I think it'll be easier and more valuable to smoke test liquid handling functionality once `dispense` is in place, too. Feel free to call me out on this if this feels wrong to you, in which case I'll come up with a smoke test plan.

### I pulled in a testing library I wrote

You may have seen my messages in Slack about [decoy](https://github.com/mcous/decoy), which I wrote because `unittest.mock` is, among other things:

- Not typed
- Heavily geared towards unconditionally-stub-then-assert-later

Both those facts make tests with mocks very difficult to read and reason with. On the JS side, #7168 pulled in [jest-when](https://github.com/timkindberg/jest-when), because jest's mock objects have the exact same problem.

This PR pulls in `decoy` to get:

- Configurable stubs
    - Fakes that only return a value if they're called with a specific set of arguments, but no-op otherwise
    - This makes mistakes like setting a return value but forgetting to check that the mock was called correctly basically impossible 
- Fully typed mocks
    - Makes test fixture type annotations way easier to read
    - Allows stub configurations (arguments and return values) to be typed, so mypy will complain if you make a mistake

See the rewrite of `test_pipetting_handler.py`, below, to get a sense of what I'm talking about

### Things I intentionally omitted

There are three items I intentionally omitted from aspirate functionality at this time:

- Aspirate in place or at an arbitrary location
    - I think the `aspirate` command should be reserved for regular, liquid-handling-from-a-well behavior
    - I think arbitrary aspirate locations (that the Protocol API supports) should be accomplished using a `move_to` command along with a (yet to be implemented) `aspirate_in_place` command, or maybe an `aspirate_at_location` command
- The `rate` parameter
    - Reasoning: rate is a little confusing (perhaps artificially so due to UI issues) and I want to make sure it's tackled in an intentional manner
- Always run `prepare_for_aspirate` if Protocol API is below 2.3
    - To be implemented in the future
    - Will be the sort of thing we add a `Config` substore or something so a Protocol Context (or whatever) can set named configuration options based on Protocol API version

These items (and others) are noted in a [Known Deviations and To Do's](https://coda.io/d/Protocol-Engine_dKsxOIJgBlG/Known-Deviations-and-To-Dos_sugyo#_luctc) page in the cookbook to ensure they are all eventually ticketed and implemented.

## Risk assessment

Non-existent to current code, because it's unused. low to medium for the eventual Protocol Engine switchover. Primary risks in that context:

- Aspirate is a pretty fundamental feature of the OT-2
- This implementation does not include all our aspirate options and features

### CODE REVIEW

File: api/src/opentrons/protocol_engine/state/pipettes.py [hunk 1]
```diff
@@ -34,10 +34,12 @@ class PipetteState:
     """Basic labware data state and getter methods."""
 
     _pipettes_by_id: Dict[str, PipetteData]
+    _aspirated_volume_by_id: Dict[str, float]
```
Comment (amitlissack) [inline api/src/opentrons/protocol_engine/state/pipettes.py:37]: An ignorant question: 
Is a single float sufficient? What about multi-channel pipettes?

Comment (mcous) [reply api/src/opentrons/protocol_engine/state/pipettes.py:37]: Cool catch. For our purposes right this second, yes this is sufficient (this is what the HardwareController does). For eventual liquid level tracking, you're right, I think this won't be sufficient. Adding it to the TODO's in the cookbook table

File: api/src/opentrons/protocol_engine/execution/pipetting.py [hunk 2]
```diff
@@ -102,3 +103,44 @@ async def drop_tip(
...
+            await self._movement_handler.move_to_well(
+                pipette_id=pipette_id,
+                labware_id=labware_id,
+                well_name=well_name,
+                well_location=WellLocation(origin=WellOrigin.TOP, offset=(0, 0, 0)),
+            )
+
+            await self._hardware.prepare_for_aspirate(mount=hw_pipette.mount)
+
+        await self._movement_handler.move_to_well(
```
Comment (mcous) [inline api/src/opentrons/protocol_engine/execution/pipetting.py:143]: Upon consideration, we need to force `move_to_well` here to issue a direct move if we run the `if not ready_to_aspirate` block. Pretty sure as written, a "prepare for aspirate" will arc to the well top, and then arc again in the movement to the aspirate location

**Review by sfoster1 — APPROVED** [review]
Looks good to me, haven't tested on a robot though.

**Review by amitlissack — APPROVED** [review]
Looks good. Love the documentation. `Decoy` looks great. I need to start adopting that.

**Pull request merged**